File: 0948-sort-an-array/solution.java

class Solution {
    public void merge(int start , int mid , int end , int[] nums){
        int i = start;
        int  j = mid + 1;
        List <Integer> ans = new ArrayList<>();
        while(i<=mid && j<=end){
            if(nums[i]<=nums[j]){
                ans.add(nums[i]);
                i++;
            } else{
                ans.add(nums[j]);
                j++;
            }
        }
        while(j<=end){
            ans.add(nums[j]);
            j++;
        }
        while(i<=mid){
            ans.add(nums[i]);
            i++;
        }
        for(int k = 0; k<ans.size(); k++){
            nums[start + k] = ans.get(k);
        }
    }
    public void mergesort(int start , int end , int[]nums){
        if(start>=end){
            return;
        }
        int mid = (start + end)/2;
        mergesort(start,mid,nums);
        mergesort(mid+1,end,nums);
        merge(start, mid, end, nums);

    }
    public int[] sortArray(int[] nums) {
        int n = nums.length;
        int start = 0;
        int end = n-1;
        mergesort(start,end,nums);
        return nums;
        
    }
}

File: 0948-sort-an-array/solution.cpp
class Solution {
public:
    void merge(int start , int mid , int end , vector<int>& nums){
        int i = start;
        int j = mid + 1;
        vector<int> ans;
        
        while(i <= mid && j <= end){
            if(nums[i] <= nums[j]){
                ans.push_back(nums[i]);
                i++;
            } else {
                ans.push_back(nums[j]);
                j++;
            }
        }
        while(i <= mid){
            ans.push_back(nums[i]);
            i++;
        }
        while(j <= end){
            ans.push_back(nums[j]);
            j++;
        }
        for(int k = 0; k < ans.size(); k++){
            nums[start + k] = ans[k];
        }
    }

    void mergesort(int start , int end , vector<int>& nums ){
        if(start >= end){
            return;
        }
        int mid = (start + end) / 2;
        mergesort(start, mid, nums);
        mergesort(mid + 1, end, nums);
        merge(start, mid, end, nums); 
    }

    vector<int> sortArray(vector<int>& nums) {
        int n = nums.size();
        mergesort(0, n - 1, nums);
        return nums;
    }
};

